File: LocalSwearFilter/run-tests.js
"use strict";

// Node-раннер тестов: `node run-tests.js`.
// Сводка — в stdout (ASCII), детали провалов — в test-output.txt (UTF-8).

var fs = require("fs");
var path = require("path");

var createProfanityFilter = require("./js/filter.js").createProfanityFilter;
var FILTER_DATA_RU = require("./js/data-ru.js");
var FILTER_DATA_EN = require("./js/data-en.js");
var FILTER_DATA_DE = require("./js/data-de.js");
var FILTER_DATA_ES = require("./js/data-es.js");
var FILTER_DATA_TR = require("./js/data-tr.js");
var FILTER_DATA_JA = require("./js/data-ja.js");
var tests = require("./js/tests.js");
var stress = require("./js/stress-corpus.js");

// Стресс-база подмешивается к основным тестам — как на странице стенда.
tests.FILTER_TESTS.push.apply(tests.FILTER_TESTS, stress.FILTER_STRESS_TESTS);

var filter = createProfanityFilter([
  FILTER_DATA_RU, FILTER_DATA_EN,
  FILTER_DATA_DE, FILTER_DATA_ES, FILTER_DATA_TR, FILTER_DATA_JA
]);
var res = tests.runFilterTests(filter);

var lines = [];
lines.push("Tests: " + res.total + ", passed: " + res.passed + ", failed: " + res.failed);
lines.push("");
for (var i = 0; i < res.results.length; i++) {
  var r = res.results[i];
  if (r.ok) continue;
  lines.push("FAIL: \"" + r.text + "\"" + (r.note ? "  (" + r.note + ")" : ""));
  lines.push("  expected profane=" + r.expected + ", got=" + r.got);
  lines.push("  censored: " + r.censored);
  var hitsStr = r.hits.map(function (h) { return h.lang + ":" + h.root; }).join(", ");
  lines.push("  hits: " + (hitsStr || "(none)"));
  lines.push("");
}

fs.writeFileSync(path.join(__dirname, "test-output.txt"), lines.join("\n"), "utf8");
console.log("tests=" + res.total + " passed=" + res.passed + " failed=" + res.failed);

// ── Бенчмарк: среднее время check() на сообщение ────────────────────────────
// Корпус — все тестовые сообщения (мат, обходы, чистые фразы вперемешку).
var corpus = tests.FILTER_TESTS.map(function (t) { return t.text; });
var warm, iter, msg;
for (warm = 0; warm < 20; warm++) for (msg = 0; msg < corpus.length; msg++) filter.check(corpus[msg]);
var REPEATS = 200;
var t0 = process.hrtime.bigint();
for (iter = 0; iter < REPEATS; iter++) {
  for (msg = 0; msg < corpus.length; msg++) filter.check(corpus[msg]);
}
var t1 = process.hrtime.bigint();
var totalMsgs = REPEATS * corpus.length;
var usPerMsg = Number(t1 - t0) / 1000 / totalMsgs;
console.log("bench: " + totalMsgs + " checks, avg " + usPerMsg.toFixed(1) + " us/message (" + (usPerMsg / 1000).toFixed(4) + " ms)");

process.exit(res.failed === 0 ? 0 : 1);

File: LocalSwearFilter/unity-dev/RunTests.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;

// Локальный раннер C#-порта (без Unity): компилируется вместе с
// ../unity/ProfanityFilter/ProfanityFilter.cs, гоняет корпус tests.txt
// и сверяет вердикты. Запуск: build-and-test.cmd
public static class RunTests
{
    public static int Main(string[] args)
    {
        string baseDir = AppDomain.CurrentDomain.BaseDirectory;
        string dataDir = Path.Combine(baseDir, @"..\unity\ProfanityFilter\Resources\ProfanityFilter");
        string[] langs = { "ru", "en", "de", "es", "tr", "ja" };
        var dataTexts = new List<string>();
        foreach (string lang in langs)
            dataTexts.Add(File.ReadAllText(Path.Combine(dataDir, lang + ".txt")));

        var filter = ProfanityFilter.FromDataTexts(dataTexts.ToArray());

        string[] lines = File.ReadAllLines(Path.Combine(baseDir, "tests.txt"));
        int total = 0, passed = 0;
        var failures = new StringBuilder();
        var corpus = new List<string>();
        foreach (string line in lines)
        {
            if (line.Length == 0) continue;
            int tab = line.IndexOf('\t');
            bool expected = line.Substring(0, tab) == "1";
            string text = line.Substring(tab + 1);
            corpus.Add(text);
            total++;
            var result = filter.Check(text);
            if (result.Profane == expected)
            {
                passed++;
            }
            else
            {
                failures.AppendLine("FAIL: \"" + text + "\" expected=" + expected + " got=" + result.Profane);
                failures.AppendLine("  censored: " + result.Censored);
            }
        }

        File.WriteAllText(Path.Combine(baseDir, "test-output-cs.txt"), failures.ToString(), Encoding.UTF8);
        Console.WriteLine("tests=" + total + " passed=" + passed + " failed=" + (total - passed));

        // бенчмарк
        for (int w = 0; w < 20; w++)
            foreach (string msg in corpus) filter.Check(msg);
        var sw = Stopwatch.StartNew();
        const int repeats = 200;
        for (int r = 0; r < repeats; r++)
            foreach (string msg in corpus) filter.Check(msg);
        sw.Stop();
        double usPerMsg = sw.Elapsed.TotalMilliseconds * 1000.0 / (repeats * corpus.Count);
        Console.WriteLine("bench: avg " + usPerMsg.ToString("F1") + " us/message");

        return total == passed ? 0 : 1;
    }
}
